File: src/oca/java/test/VMGroupTest.java
/*******************************************************************************
 * Copyright 2002-2016, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
import static org.junit.Assert.assertTrue;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.group.Group;
import org.opennebula.client.vmgroup.*;
import org.opennebula.client.user.User;




public class VMGroupTest
{

    private static VMGroup vmg;
    private static VMGroupPool pool;

    private static Client client;

    private static OneResponse res;
    private static String name = "new_test_vmg";


    private static String template_str =
        "NAME = \"" + name + "\"\n" +
        "DESCRIPTION  = \"test vmgroup\"\n"+
        "ATT1 = \"VAL1\"\n" +
        "ATT2 = \"VAL2\"";

    /**
     * @throws java.lang.Exception
     */
    @BeforeClass
    public static void setUpBeforeClass() throws Exception
    {
        client  = new Client();
        pool    = new VMGroupPool(client);
    }

    /**
     * @throws java.lang.Exception
     */
    @AfterClass
    public static void tearDownAfterClass() throws Exception
    {
    }

    /**
     * @throws java.lang.Exception
     */
    @Before
    public void setUp() throws Exception
    {
        res = VMGroup.allocate(client, template_str);

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        vmg = new VMGroup(oid, client);
    }

    /**
     * @throws java.lang.Exception
     */
    @After
    public void tearDown() throws Exception
    {
        vmg.delete();
    }

    @Test
    public void allocate()
    {
        vmg.delete();

        res = VMGroup.allocate(client, template_str);
        assertTrue( res.getErrorMessage(), !res.isError() );

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        vmg = new VMGroup(oid, client);


        pool.info();

        boolean found = false;
        for(VMGroup temp : pool)
        {
            found = found || temp.getName().equals(name);
        }

        assertTrue( found );
    }

    @Test
    public void info()
    {
        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.getName().equals(name) );
    }

    @Test
    public void update()
    {
        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.xpath("TEMPLATE/ATT1").equals( "VAL1" ) );
        assertTrue( vmg.xpath("TEMPLATE/ATT2").equals( "VAL2" ) );

        String new_vmg =   "ATT2 = NEW_VAL\n" +
                                "ATT3 = VAL3";

        res = vmg.update(new_vmg);
        assertTrue( res.getErrorMessage(), !res.isError() );


        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );
        assertTrue( vmg.xpath("TEMPLATE/ATT1").equals( "" ) );
        assertTrue( vmg.xpath("TEMPLATE/ATT2").equals( "NEW_VAL" ) );
        assertTrue( vmg.xpath("TEMPLATE/ATT3").equals( "VAL3" ) );
    }

    @Test
    public void chmod()
    {
        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        String owner_a = vmg.xpath("PERMISSIONS/OWNER_A");
        String group_a = vmg.xpath("PERMISSIONS/GROUP_A");

        res = vmg.chmod(0, 1, -1, 1, 0, -1, 1, 1, 0);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_A").equals(owner_a) );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_A").equals(group_a) );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_A").equals("0") );
    }

    @Test
    public void chmod_octet()
    {
        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.chmod(640);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.xpath("PERMISSIONS/OWNER_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_A").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_U").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_M").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_A").equals("0") );

        res = vmg.chmod("147");
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_M").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OWNER_A").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( vmg.xpath("PERMISSIONS/OTHER_A").equals("1") );
    }

    @Test
    public void attributes()
    {
        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

//        assertTrue( vmg.xpath("ID").equals("0") );
        assertTrue( vmg.xpath("NAME").equals(name) );
    }

    @Test
    public void delete()
    {
        res = vmg.delete();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.info();
        assertTrue( res.isError() );
    }

    @Test
    public void chown()
    {
        // Create a new User and Group
        res = User.allocate(client, "vmg_test_user", "password");
        assertTrue( res.getErrorMessage(), !res.isError() );

        int uid = Integer.parseInt(res.getMessage());

        res = Group.allocate(client, "vmg_test_group");
        assertTrue( res.getErrorMessage(), !res.isError() );

        int gid = Integer.parseInt(res.getMessage());

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.uid() == 0 );
        assertTrue( vmg.gid() == 0 );

        res = vmg.chown(uid, gid);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.uid() == uid );
        assertTrue( vmg.gid() == gid );

        res = vmg.chgrp(0);

        res = vmg.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmg.uid() == uid );
        assertTrue( vmg.gid() == 0 );
    }
}

File: src/oca/go/src/goca/vmgroup_test.go
// +build !disabled

/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	//"testing"
	//"strconv"

	//ds "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/datastore"
	//dskeys "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/datastore/keys"
	//"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vmgroup"
	//"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm/keys"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
	. "gopkg.in/check.v1"
)

// Hook up gocheck into the "go test" runner.
//func Test(t *testing.T) { TestingT(t) }

type VMGroupSuite struct {
	vmgID int
}

var _ = Suite(&VMGroupSuite{})

func (s *VMGroupSuite) SetUpSuite(c *C) {
}

func (s *VMGroupSuite) SetUpTest(c *C) {
	vmgTmpl := `NAME = "VM Group Test"
		ROLE = [ NAME = web ]
		ROLE = [ NAME = db, POLICY = "AFFINED"  ]
		ROLE = [ NAME = app  ]
		ROLE = [ NAME = bck  ]

		AFFINED = "web, bck"
		ANTI_AFFINED = "db"
		`
	vmgID, err := testCtrl.VMGroups().Create(vmgTmpl)
	c.Assert(err, IsNil)
	s.vmgID = vmgID
}

func (s *VMGroupSuite) TearDownTest(c *C) {
	// Delete VM Group
	vmgC := testCtrl.VMGroup(s.vmgID)
	vmgC.Delete()
}

func (s *VMGroupSuite) TearDownSuite(c *C) {
}

////////////////////////////////////////////////////////////////////////////////

func (s *VMGroupSuite) TestCreated(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Name, Equals, "VM Group Test")

	// Test roles
	c.Assert(vmg.Roles[0].ID, Equals, 0)
	c.Assert(vmg.Roles[0].Name, Equals, "web")
	c.Assert(vmg.Roles[1].ID, Equals, 1)
	c.Assert(vmg.Roles[1].Name, Equals, "db")
	c.Assert(vmg.Roles[1].Policy, Equals, "AFFINED")
	c.Assert(vmg.Roles[2].ID, Equals, 2)
	c.Assert(vmg.Roles[2].Name, Equals, "app")
	c.Assert(vmg.Roles[3].ID, Equals, 3)
	c.Assert(vmg.Roles[3].Name, Equals, "bck")

	// Get Backup Job by Name
	id, err := testCtrl.VMGroups().ByName(vmg.Name)
	c.Assert(err, IsNil)
	c.Assert(id, Equals, s.vmgID)

}

func (s *VMGroupSuite) TestUpdate(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.Update("A=B", 1)
	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	val, _ := vmg.Template.GetStr("A")
	c.Assert(val, Equals, "B")
}

func (s *VMGroupSuite) TestChmod(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Permissions.String(), Equals, "um-------")

	changedPerm := shared.Permissions{1, 1, 1, 1, 1, 1, 1, 1, 1}
	vmgC.Chmod(changedPerm)

	vmg, err = vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(*vmg.Permissions, Equals, changedPerm)
}

func (s *VMGroupSuite) TestChown(c *C) {
	// Test only if the call exists, no real change
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.Chown(1, 1)

	c.Assert(err, IsNil)
}

func (s *VMGroupSuite) TestRename(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.Rename("Renamed")
	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Name, Equals, "Renamed")
}

func (s *VMGroupSuite) TestLock(c *C) {
	// Lock
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.Lock(shared.LockUse)

	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)

	c.Assert(err, IsNil)
	c.Assert(vmg.LockInfos.Locked, Equals, 1);

	// Unlock
	err = vmgC.Unlock()
	c.Assert(err, IsNil)

	vmg, err = vmgC.Info(false)

	c.Assert(err, IsNil)
	c.Assert(vmg.LockInfos, IsNil)
}

func (s *VMGroupSuite) TestRoleAdd(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.RoleAdd(`ROLE = [ name = "other" ]`)
	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Roles[4].ID, Equals, 4)
	c.Assert(vmg.Roles[4].Name, Equals, "other")
}

func (s *VMGroupSuite) TestRoleDelete(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.RoleDelete(2) // app
	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Roles[2].ID, Equals, 3)
	c.Assert(vmg.Roles[2].Name, Equals, "bck")
}

func (s *VMGroupSuite) TestRoleUpdate(c *C) {
	vmgC := testCtrl.VMGroup(s.vmgID)
	err := vmgC.RoleUpdate(2, `ROLE = [ POLICY = "ANTI_AFFINED" ]`)
	c.Assert(err, IsNil)

	vmg, err := vmgC.Info(false)
	c.Assert(err, IsNil)

	c.Assert(vmg.Roles[2].Name, Equals, "app")
	c.Assert(vmg.Roles[2].Policy, Equals, "ANTI_AFFINED")
}
